### ISSUE TITLE
Enhancement: no-unsafe-function-type should allow `Extract<Type, { fn: Function }>`

### ISSUE THREAD

**Issue by franklin-ross** [typescript-eslint/typescript-eslint#9831]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-unsafe-function-type/

### Description

I propose this rule should allow `Function` in the second argument of `Extract<>`, and possibly others, because `Function` doesn't end up in the output in that case.

### Fail

```TypeScript
// No new failures
```

### Pass

```TypeScript
type A { foo: int };
type B { foo(): number }
type C { foo(str: string): void }

type All = A | B | C

// In this case, `x` is type `B | C`
function doThings(x: Extract<All , { foo: Function }>): void { ... }
```

### Additional Info

I like this because it's a concise and self-documenting way to select the types I need out of a union based on the needs of the code I'm writing and will automatically update if `All` changes compared to manually listing the types I want.

**Comment by Josh-Cena** [comment]
The problem with this is that it's secretly unsafe. If you have:

```ts
class X {}
type D = { foo: typeof X };
```

Do you expect `D` to be present in the union, despite it not being actually callable? Even worse, what about:

```ts
type E = {
    foo: { bind: any; call: any; apply: any; prototype: any; length: any; arguments: any; caller: any; name: any; [Symbol.hasInstance]: any }
}
```

As mentioned, `Function` is a structural type and there's actually no guarantee that it's a function.

`Extract<All, { foo: (...args: never) => unknown }>` is so much safer and not much less readable.

On the more technical side, this is impossible to generalize as a rule—similar to other rules that blanket-ban types, including `no-explicit-any`, we are purposed not trying to be smart and force you to think whether it's safe and whether compliant alternatives exist. If you just like the current style and accept the tradeoffs, a disable comment is the way to go.

**Comment by bradzacher** [comment]
Well said Josh 😄

Here's a [playground](https://www.typescriptlang.org/play/?#code/C4TwDgpgBAglC8UDeUBmB7dAuKA7ArgLYBGEATlAL4DcAUKJFAEILJqYAUAlDgSeVToNoAYVYoM6DgGdgZHLLIBLXAHMeUAG7olAE0G0AxgBsAhtOlQAGskr1w0ACLj22KMPSprB4VACi4rRQwa44KMQqujimuCDUUIamxsbRsfGmYGDGIKlxUGBk6MBFDrnxxhBqwAAWZVCmZKpElcDSdYnJ5HW4poQQdQDaAMogJOjGAHTV5gCSuLIxhhAAurkGNLT2jDDJrHAAPsxQh2KHzod+dLQA9NdQc+7VSpaJ0hAANFAABgAeX1DPdwOb4sE7HKDnfxfWiofC4QzAJToXBQXToAAqTzU0gAqgA5IYwABifg4Pxwfh+clMCIAPDtjJ8JJgcES4QikSjKAA+DTaPTIIIhW4hUVi8UAPQA-LQ7JsRQ8aoDXh9vn8AZZfF9QVARNDYfDEcjURisappISSWSKVSyDTgPTkkzQlAOBN3Q1zbwIJpyFwENyoHCANa4dAAdy5vJw-P0SCFwRF4uTwWlsqAA) with Josh's examples.

`Function` is UNSAFE, no matter the context.

If you want to use it in this case and accept the unsafety -- that's acceptable! You just need a disable comment.
